### ISSUE TITLE
Postinstall script installation fails when installing locally

### ISSUE THREAD

**Issue by horike37** [serverless/serverless#3792]
# This is a Bug Report

## Description
I know a correct way to install globally using `npm install serverless - g`, but when installing locally in own project with `npm install serverless --save`, an error occurs during execution of postinstall.js.

Is this bug? or does it mean sls has not supported for installing locally?

### step to reproduce
- Create directory with `mkdir sls-project && cd sls-project`
- Run `npm init`
- Run `npm install serverless --save`, and you will see the following error

```
$npm install serverless --save
npm WARN registry Unexpected warning for https://registry.npmjs.org/: Miscellaneous Warning EINTEGRITY: sha1-F2n0tVHu3OQZ8FBd6uLiZ2NULTc= integrity checksum failed when using sha1: wanted sha1-F2n0tVHu3OQZ8FBd6uLiZ2NULTc= but got sha512-0TZ20a+xcIl4u0+Mj5xDH2yOWdmQiXlKf9Hm+TgDXjTMsEYb+gDrmb8e8UNAzMCitX8NBqG4Z/FUQIyzv/R1JQ==. (2982 bytes)

> serverless@1.15.3 postinstall /Users/horike/src/sls-project/node_modules/serverless
> node ./scripts/postinstall.js

module.js:471
    throw err;
    ^

Error: Cannot find module '/Users/horike/src/sls-project/node_modules/serverless/node_modules/tabtab/src/cli.js'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
module.js:471
    throw err;
    ^

Error: Cannot find module '/Users/horike/src/sls-project/node_modules/serverless/node_modules/tabtab/src/cli.js'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3

> spawn-sync@1.0.15 postinstall /Users/horike/src/sls-project/node_modules/spawn-sync
> node postinstall

npm WARN saveError ENOENT: no such file or directory, open '/Users/horike/src/sls-project/package.json'
npm notice created a lockfile as package-lock.json. You should commit this file.
npm WARN enoent ENOENT: no such file or directory, open '/Users/horike/src/sls-project/package.json'
npm WARN sls-project No description
npm WARN sls-project No repository field.
npm WARN sls-project No README data
npm WARN sls-project No license field.

+ serverless@1.15.3
added 210 packages in 23.298s
```
Similar or dependent issues:
* #3760
* #3763

## Additional Data
* ***node version***: 6.10.2
* ***npm version***: 5.0.3
* ***Serverless Framework Version you're using***:1.15.3
* ***Operating System***: macOS 10.11.6
* ***Stack Trace***: above
* ***Provider Error messages***: above

**Comment by HyperBrain** [comment]
I think I know what that is - @horike37 can you check if the tabtab package in your project is located at `node_modules/tabtab` instead of `node_modules/serverless/tabtab`.

This can happen is npm flattens the module directory and optimizes it.
@pmuens If that's the case, the postinstall call of the tabtab cli should try to find the cli.js in one of the locations and execute it from where it is actually located.

**Comment by horike37** [comment]
@HyperBrain 
I think what you stated correctly.
When instaling globally, the tabtab package is located at `serverless/node_modules/tabtab/src/cli.js`, so `sls` can run the cli.
However, when instaling in my project, the cli is located at `./node_modules/tabtab/src/cli.js`, so `sls` can not find the cli. `sls` has expected the cli installed at `./node_modules/serverless/node_modules/tabtab/src/cli.js`

**Comment by HyperBrain** [comment]
as a quick fix you can use `npm install --legacy-bundling` in your project dir. This will make sure that npm disables the optimization and flattening.

**Comment by horike37** [comment]
>as a quick fix you can use npm install --legacy-bundling in your project dir
That's make sense! I didn't know it

**Comment by HyperBrain** [comment]
@pmuens I use the `get-installed-path` node package in some of my code. This let's you find the installation location of node modules easily. You could use that to retrieve the current location of the tabtab module:

```
const getInstalledPath = require('get-installed-path');
...
const tabtabPath = getInstalledPath.sync('tabtab', { local: true });
```

**Comment by k1LoW** [comment]
@HyperBrain Thank you for your advice!

**Comment by pmuens** [comment]
Thanks for jumping in and debugging @HyperBrain 👍 

We'll look into this! /cc @eahefnawy @brian

**Comment by hash** [comment]
Hello, on windows it is still broken. 
My fix attempt:
```diff
diff --git a/scripts/postinstall.js b/scripts/postinstall.js
index 9eb4fba3..5de81cc0 100644
--- a/scripts/postinstall.js
+++ b/scripts/postinstall.js
@@ -5,6 +5,7 @@
 
 const Serverless = require('../lib/Serverless');
 const execSync = require('child_process').execSync;
+const path = require('path');
 
 try {
   const serverless = new Serverless();
@@ -20,14 +21,16 @@ try {
 
 function setupAutocomplete() {
   return new Promise((resolve, reject) => {
-    const tabtabPath = require.resolve('tabtab').replace(/\/index.js$/, '');
+    const tabtabPath = require.resolve('tabtab').split(path.sep);
+    if (tabtabPath[tabtabPath.length - 1] === 'index.js') { tabtabPath.pop(); }
+    const tabtabPathCli = path.join(...tabtabPath, 'src', 'cli.js');
     try {
-      execSync(`node ${tabtabPath}/src/cli.js install --name serverless --auto`);
-      execSync(`node ${tabtabPath}/src/cli.js install --name sls --auto`);
+      execSync(`node ${tabtabPathCli} install --name serverless --auto`);
+      execSync(`node ${tabtabPathCli} install --name sls --auto`);
       return resolve();
     } catch (error) {
-      execSync(`node ${tabtabPath}/src/cli.js install --name serverless --stdout`);
-      execSync(`node ${tabtabPath}/src/cli.js install --name sls --stdout`);
+      execSync(`node ${tabtabPathCli} install --name serverless --stdout`);
+      execSync(`node ${tabtabPathCli} install --name sls --stdout`);
       console.log('Could not auto-install serverless autocomplete script.');
       console.log('Please copy / paste the script above into your shell.');
       return reject(error);

```

**Comment by pmuens** [comment]
Hey @hash thanks for pointin that out.  I just PRed a fix in #3870.

Would be nice if you could try it out. Thanks 👍

**Comment by muxahuk** [comment]
`
C:\Program Files\nodejs\node_modules\serverless>npm install

> serverless@1.16.1 postinstall C:\Program Files\nodejs\node_modules\serverless
> node ./scripts/postinstall.js

module.js:487
    throw err;
    ^

Error: Cannot find module 'C:\Users\M7\AppData\Roaming\nvm\v8.1.3\node_modules\serverless\node_modules\tabtab\index.js\src\cli.js'
    at Function.Module._resolveFilename (module.js:485:15)
    at Function.Module._load (module.js:437:25)
    at Function.Module.runMain (module.js:605:10)
    at startup (bootstrap_node.js:158:16)
    at bootstrap_node.js:575:3
module.js:487
    throw err;
    ^

Error: Cannot find module 'C:\Users\M7\AppData\Roaming\nvm\v8.1.3\node_modules\serverless\node_modules\tabtab\index.js\src\cli.js'
    at Function.Module._resolveFilename (module.js:485:15)
    at Function.Module._load (module.js:437:25)
    at Function.Module.runMain (module.js:605:10)
    at startup (bootstrap_node.js:158:16)
    at bootstrap_node.js:575:3
up to date in 4.391s
`
fixed with:
`const tabtabPath = require.resolve('tabtab').replace(/(\/|\\)index.js$/, '');`

**Comment by pmuens** [comment]
@muxahuk thanks for your comment and for providing the fix for it.

It looks like you've installed the release from `npm` which does not include the fix https://github.com/serverless/serverless/pull/3870 yet (it was merged into `master` yesterday).

Could you try to re-install from the `master` in GitHub and see if it's fixed?

**Comment by muxahuk** [comment]
@pmuens Yes, it works with master.

**Comment by pmuens** [comment]
> @pmuens Yes, it works with master.

Phew! Thanks for confirming @muxahuk 👍 💯

**Comment by shawnmclean** [comment]
`npm install -g serverless/serverless` should work for now for anyone else looking.

**Comment by JimiHFord** [comment]
I got this error with `npm v5.6.0` when I changed from a global install of serverless, to a local install. Clearing the `node_modules` cache on my CI box solved the problem for me (and I was able to keep the local install of `serverless v1.27.3`

**Comment by EliaContini** [comment]
Hi,

I have a similar issue. I'm trying to install Serverless locally on Ubuntu 18.04.5 LTS.
Node.js: 14.16.0
npm: 7.5.6
Node.js and npm installed as snap

Steps to reproduce
1. create a folder (mkdir test && cd test)
2. npm init
3. npm install serverless (I tried even with sudo npm install -g serverless)

I got this error

```
npm ERR! code 1
npm ERR! path /home/elia/projects/test/node_modules/serverless
npm ERR! command failed
npm ERR! command sh -c node ./scripts/postinstall.js
```
No ```node_modules``` folder is created during the process.

I even tried to delete npm cache as suggested in a [forum post](https://forum.serverless.com/t/error-installing-serverless-using-npm-on-ubuntu-20-04-desktop/11297) but did not work.
